model/CSVDialog.cc

```diff
@@ -34,6 +53,111 @@ void CSVDialog::reportFromFile(const std::string& input, const std::string& outp
 }
 
+// Return file name after downloading a CSV file from the web.
+std::string CSVDialog::loadWebFile(const std::string& url)
+{
+    try
+    {
+		// Parse input URL
+        boost::regex ex("(http|https)://([^/ :]+):?([^/ ]*)(/?[^ #?]*)\\x3f?([^ #]*)#?([^ ]*)");
+        boost::cmatch what;
+        regex_match(url.c_str(), what, ex);
+        
+        // The web domain 
+        auto const host = what[2];
+        
+        // Set the port type
+        string port = string(what[1].first, what[1].second);
+        if (string(what[1].first, what[1].second)=="https") port = "https"; 
+        else port = "http";
+        auto const target = string(what[4].first, what[4].second);  
+        
+        // Protocol version
+        int version;
+        if (!string(what[6].first, what[6].second).empty() && !std::strcmp("1.0", what[6].second)) version= 11;
+        else version=10;
+		
+        // The io_context is required for all I/O
+        boost::asio::io_context ioc;
+        
+        // The SSL context is required, and holds certificates
+        ssl::context ctx{ssl::context::tls_client};
+        
+        // Verify the remote server's certificate
+        ctx.set_verify_mode(ssl::verify_peer | ssl::context::verify_fail_if_no_peer_cert);    
+        ctx.set_default_verify_paths();
+        
+        // tag::ctx_setup_source[]
+        boost::certify::enable_native_https_server_verification(ctx);
+        // end::ctx_setup_source[]        
+        		        
+        // These objects perform our I/O
+        tcp::resolver resolver{ioc};
+        ssl::stream<tcp::socket> stream{ioc, ctx};        
+        
+        // tag::stream_setup_source[]
+        boost::certify::set_server_hostname(stream, string(what[2].first, what[2].second));
+        boost::certify::sni_hostname(stream, host);
+        // end::stream_setup_source[]         
+
+        // Look up the domain name
+        auto const results = resolver.resolve(host, what[1]);
+                
+        // Make the connection on the IP address we get from a lookup
+        boost::asio::connect(stream.next_layer(), results.begin(), results.end());                   
+             
+        // Perform the SSL handshake
+        stream.handshake(ssl::stream_base::client);             
+
+        // Set up an HTTP GET request message
+        http::request<http::string_body> req{http::verb::get, target, version};
+        req.set(http::field::host, host);
+        req.set(http::field::user_agent, BOOST_BEAST_VERSION_STRING);
+
+        // Send the HTTP request to the remote host
+        http::write(stream, req);
+
+        // This buffer is used for reading and must be persisted
+        boost::beast::flat_buffer buffer;
+
+        // Declare a container to hold the response
+        http::response_parser<http::string_body> res;
+        res.body_limit((std::numeric_limits<std::uint64_t>::max)());
+
+        // Receive the HTTP response
+        http::read(stream, buffer, res);
```

Comment:
You need to check the response code. Anything other than 200 OK should result in an exception being thrown.

Reply:
I think - need to check what happens in the case of a redirect.

Reply:
I think I've fixed this now.